## Pembuatan Hipotesis

Tujuan Projek: Mencari tahu pengaruh status perkawinan seorang nasabah, jumlah anak yang dimiliki, pendapatan, dan perbedaan tujuan dalam meminjam terhadap probabilitas gagal bayar dalam melunasi pinjaman.

Tujuan membuktikan apakah:
1. Seseorang gagal dalam membayar pinjaman karena ada pengeluaran lain untuk istri dan anaknya.
2. Seseorang gagal dalam membayar pinjaman karena pendapatannya yang ternyata tidak cukup.
3. Perbedaan tujuan dalam melakukan pinjaman bisa menjadi faktor seseorang gagal dalam melakukan pinjaman.

In [1]:
import pandas as pd

try:
    df = pd.read_csv('D:\datasets\credit_scoring_eng.csv')
except:
    df = pd.read_csv('/datasets/credit_scoring_eng.csv') 

In [2]:
df.shape

(21525, 12)

Data ini memiliki baris sebanyak 21525 dan kolom sebanyak 12

In [3]:
df.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42    bachelor's degree             0   
1          1   -4024.803754         36  secondary education             1   
2          0   -5623.422610         33  Secondary Education             1   
3          3   -4124.747207         32  secondary education             1   
4          0  340266.072047         53  secondary education             1   
5          0    -926.185831         27    bachelor's degree             0   
6          0   -2879.202052         43    bachelor's degree             0   
7          0    -152.779569         50  SECONDARY EDUCATION             1   
8          2   -6929.865299         35    BACHELOR'S DEGREE             0   
9          0   -2188.756445         41  secondary education             1   
10         2   -4171.483647         36    bachelor's degree             0   
11         0    -792.701887         40  secondary education             1   
12         0            NaN         65  secondary education             1   
13         0   -1846.641941         54         some college             2   
14         0   -1844.956182         56    bachelor's degree             0   
15         1    -972.364419         26  secondary education             1   
16         0   -1719.934226         35  secondary education             1   
17         0   -2369.999720         33    bachelor's degree             0   
18         0  400281.136913         53  secondary education             1   
19         0  -10038.818549         48  SECONDARY EDUCATION             1   

        family_status  family_status_id gender income_type  debt  \
0             married                 0      F    employee     0   
1             married                 0      F    employee     0   
2             married                 0      M    employee     0   
3             married                 0      M    employee     0   
4   civil partnership                 1      F     retiree     0   
5   civil partnership                 1      M    business     0   
6             married                 0      F    business     0   
7             married                 0      M    employee     0   
8   civil partnership                 1      F    employee     0   
9             married                 0      M    employee     0   
10            married                 0      M    business     0   
11            married                 0      F    employee     0   
12  civil partnership                 1      M     retiree     0   
13            married                 0      F    employee     0   
14  civil partnership                 1      F    business     1   
15            married                 0      F    employee     0   
16            married                 0      F    employee     0   
17  civil partnership                 1      M    employee     0   
18    widow / widower                 2      F     retiree     0   
19           divorced                 3      F    employee     0   

    total_income                              purpose  
0      40620.102                purchase of the house  
1      17932.802                         car purchase  
2      23341.752                purchase of the house  
3      42820.568              supplementary education  
4      25378.572                    to have a wedding  
5      40922.170                purchase of the house  
6      38484.156                 housing transactions  
7      21731.829                            education  
8      15337.093                     having a wedding  
9      23108.150  purchase of the house for my family  
10     18230.959                      buy real estate  
11     12331.077           buy commercial real estate  
12           NaN                    to have a wedding  
13     20873.317                         car purchase  
14     26420.466          buy residential real estate  
15     18691.345         construction of own property  
16   

**Hal yang dilihat:**

1. Untuk kolom days_employed memiliki nilai yang tidak sesuai yaitu nilai negatif. Yang mana, pada kolom tersebut dicantumkan lama pekerja dalam hari yang seharusnya tidak negatif. 
2. Untuk kolom education, ada baris yang bertuliskan dengan huruf kapital. Hal ini bisa memengaruhi analisis karena program akan melihat ini sebagai hal yang berbeda.
3. Terdapat nilai yang hilang di kolom days_employed dan total_income.

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


**Hal yang saya lihat:**
1. Pada kolom days_employed dan total_income terdapat data yang hilang. 
2. Hal ini bisa memengaruhi proses analisis. 
3. Hal ini juga mungkin bisa terjadi karena ada orang yang sudah pensiun tapi melakukan peminjaman. 
4. Munkgin di kolom days_employed juga harus diubah tipe datanya menjadi int64 karena jumlah hari kerja tidak seharusnya bernilai koma.

In [5]:
df[df['days_employed'].isnull()]

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

Jika dilihat satu baris nilai yang hilang pada kolom days_employed memiliki nilai yang sama pada baris total_income. Ini bisa menjadi asumsi bahwa ada orang yang sudah pensiun atau ada kesalahan dalam input data.

Ada 2174 nilai yang hilang pada kolom days_employed dan total_income. Hal ini sesuai dengan tabel sebelumnya.

In [6]:
df.loc[(df['days_employed'].isna()) & (df['total_income'].isna())].shape

(2174, 12)

Jumlah baris yang ada pada kedua kolom tersebut memiliki jumlah yang sama.

**Kesimpulan Sementara**

Jumlah baris keseluruhan data ada 21525 dan untuk jumlah baris data yang hilang ada 2174. Beberapa kolom mendeskripsikan tentang bagaimana kondisi dari si peminjam: children, family_status, dan total_income. Namun, ada beberapa nilai yang hilang di kolom yang penting yaitu total_income. 

Karakteristik yang menyebabkan data tersebut hilang bisa jadi karena pekerjaan dan usia seseorang. Karena bisa jadi karena ada yang sudah pensiun, mereka tidak lagi perlu mengisi nilai pada bagian dob_years dan total_income. Nilai yang hilang tersebut memiliki ketergantungan dengan karakteristik yang ditentukan yaitu dob_years dan income_type.

In [7]:
percantage_of_days_employed = df['days_employed'].isna().sum() * 100/len(df)
percantage_of_total_income = df['total_income'].isna().sum() * 100/len(df)
print(f'Presentasi data yang hilang adalah pada kolom days_employed: {percantage_of_days_employed}')
print(f'Presentasi data yang hilang adalah pada kolom total_income: {percantage_of_total_income}')

Presentasi data yang hilang adalah pada kolom days_employed: 10.099883855981417
Presentasi data yang hilang adalah pada kolom total_income: 10.099883855981417


Presentase data yang hilang ada 10%. Presentasi yang tidak terlalu besar tapi akan berpengaruh terhadap analisis kedepannya, karena data yang hilang ada di kolom total_income. Karena salah satu faktor kita mellihat seseorang bisa membayar pinjaman bisa dilihat dari pendapatan seseorang. Maka data yang hilang tersebut perlu diisi dengan nilai mean atau median.

Langkah selanjutnya adalah mengisi nilai yang hilang tapi sebelum itu saya harus memperbaiki isi data seperti pada kolom children, dob_years, dan education. 

In [8]:
df_null = df[(df['days_employed'].isnull()) & (df['total_income'].isnull())]
df_null

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

In [9]:
df_null[df_null['income_type']=='civil servant']

children  days_employed  dob_years            education  education_id  \
26            0            NaN         41  secondary education             1   
41            0            NaN         50  secondary education             1   
72            1            NaN         32    bachelor's degree             0   
242           0            NaN         58  secondary education             1   
389           1            NaN         31  SECONDARY EDUCATION             1   
...         ...            ...        ...                  ...           ...   
20469         0            NaN         57  secondary education             1   
20479         1            NaN         26    bachelor's degree             0   
20914         0            NaN         32    bachelor's degree             0   
21242         1            NaN         33    bachelor's degree             0   
21268         1            NaN         44    bachelor's degree             0   

           family_status  family_status_id gender    income_type  debt  \
26               married                 0      M  civil servant     0   
41               married                 0      F  civil servant     0   
72               married                 0      M  civil servant     0   
242              married                 0      F  civil servant     0   
389             divorced                 3      M  civil servant     0   
...                  ...               ...    ...            ...   ...   
20469    widow / widower                 2      M  civil servant     1   
20479            married                 0      F  civil servant     0   
20914            married                 0      F  civil servant     0   
21242          unmarried                 4      F  civil servant     0   
21268  civil partnership                 1      F  civil servant     0   

       total_income                                   purpose  
26              NaN                                 education  
41              NaN                  second-hand car purchase  
72              NaN  transactions with commercial real estate  
242             NaN                  purchase of my own house  
389             NaN                   supplementary education  
...             ...                                       ...  
20469           NaN                              to own a car  
20479           NaN                                 education  
20914           NaN                              car purchase  
21242           NaN              construction of own property  
21268           NaN                          having a wedding  

[147 rows x 12 columns]

In [10]:
percentage = df_null['income_type'].value_counts(normalize=True).reset_index().rename(columns={"income_type":"percentage"})
percentage['percentage'] = percentage['percentage'].apply("{:,.2%}".format)
count = df_null['income_type'].value_counts().reset_index().rename(columns={"income_type":"counts"})
distribution = pd.concat([percentage,count[['counts']]], axis=1)
distribution

index percentage  counts
0       employee     50.83%    1105
1       business     23.37%     508
2        retiree     19.00%     413
3  civil servant      6.76%     147
4   entrepreneur      0.05%       1

Pekerjaan yang memiliki paling banyak nilai hilang adalah employee. Salah satunya penyebabnya mungkin ada beberapa yang sudah penisun dan sudah tidak perlu mengisi total_income.

**Kemungkinan penyebab hilangnya nilai dalam data**

Terdapat pola tertentu dalam hilangnya data. Setiap data di kolom days_employed hilang, maka data di kolom total_income juga ikut hilang. Dan hal itu banyak terjadi di baris yang sama. Mungkin mereka ada yang sudah pensiun, belum mendapatkan pekerjaan, atau baru mendapatkan pekerjaan sehingga masih belum ada pemasukan. Namun, tidak menutup kemungkinan juga mereka hanya ingin tidak mencamtukan income dan lama mereka bekerja.

In [11]:
df['income_type'].value_counts(normalize=True)

employee                       0.516562
business                       0.236237
retiree                        0.179141
civil servant                  0.067782
unemployed                     0.000093
entrepreneur                   0.000093
student                        0.000046
paternity / maternity leave    0.000046
Name: income_type, dtype: float64

**Kesimpulan Sementara**

Setelah diperiksa ternyata dataset keseluruhan dengan dataset yang difilter memiliki nilai distribusi yang sama. Ini berarti data yang hilang terjadi secara random hampir di setiap data. Hal ini mematahkan asumsi awal kalau data yang hiang di total_income terjadi karena jenis pekerjaan nasabah. Bisa jadi terjadi karena kesalahan dalam input atau terjadi error.

In [12]:
df_null_pivot = df_null.pivot_table(index='dob_years', columns='income_type', values='debt', aggfunc='count')
df_null_pivot

income_type  business  civil servant  employee  entrepreneur  retiree
dob_years                                                            
0                 2.0            NaN       5.0           NaN      3.0
19                NaN            NaN       1.0           NaN      NaN
20                1.0            NaN       4.0           NaN      NaN
21                7.0            1.0      10.0           NaN      NaN
22                6.0            NaN      11.0           NaN      NaN
23                5.0            1.0      30.0           NaN      NaN
24                9.0            1.0      10.0           NaN      1.0
25                4.0            4.0      15.0           NaN      NaN
26                9.0            2.0      24.0           NaN      NaN
27                6.0            3.0      27.0           NaN      NaN
28               22.0            4.0      31.0           NaN      NaN
29               13.0            6.0      31.0           NaN      NaN
30               10.0            4.0      44.0           NaN      NaN
31               28.0            7.0      29.0           NaN      1.0
32                8.0            4.0      25.0           NaN      NaN
33               12.0            7.0      32.0           NaN      NaN
34               22.0            4.0      43.0           NaN      NaN
35               25.0            2.0      37.0           NaN      NaN
36               16.0            9.0      36.0           NaN      2.0
37               16.0            2.0      35.0           NaN      NaN
38               21.0            3.0      29.0           NaN      1.0
39               10.0            6.0      34.0           NaN      1.0
40               26.0            9.0      30.0           NaN      1.0
41               14.0            2.0      42.0           NaN      1.0
42               20.0            1.0      40.0           NaN      4.0
43               13.0            4.0      33.0           NaN      NaN
44                8.0            3.0      33.0           NaN      NaN
45               13.0            2.0      32.0           NaN      3.0
46                8.0            5.0      34.0           NaN      1.0
47               16.0            5.0      38.0           NaN      NaN
48               12.0            7.0      24.0           NaN      3.0
49                9.0            3.0      36.0           NaN      2.0
50               15.0            4.0      27.0           NaN      5.0
51               17.0            3.0      20.0           NaN     10.0
52               16.0            5.0      21.0           NaN     11.0
53                8.0            2.0      19.0           NaN     15.0
54               13.0            2.0      22.0           NaN     18.0
55               15.0            3.0      15.0           NaN     15.0
56                7.0            4.0      18.0           NaN     25.0
57                4.0            2.0      13.0           NaN     37.0
58                6.0            3.0      14.0           1.0     32.0
59                4.0            1.0       7.0           NaN     22.0
60                4.0            2.0       6.0           NaN     27.0
61                2.0            NaN       8.0           NaN     28.0
62                3.0            1.0       6.0           NaN     28.0
63                NaN            1.0      10.0           NaN     18.0
64                NaN            3.0       5.0           NaN     29.0
65                NaN            NaN       3.0           NaN     17.0
66                1.0            NaN       1.0           NaN     18.0
67                NaN            NaN       1.0           NaN     15.0
68                NaN            NaN       2.0           NaN      7.0
69                1.0            NaN       1.0           NaN      3.0
70                NaN            NaN       1.0           NaN      2.0
71                NaN            NaN       NaN           NaN      5.0
72                1.0            NaN       NaN           NaN      1.0
73        

**Kesimpulan Sementara**

Nilainya cukup random, tidak ada pola unik yang dijadikan alasan mengapa data itu bisa hilang. Karena dari table diatas mengatakan kalau data yang hilang bisa terjadi dari semua jenis pekerjaan dan semua umur.

In [13]:
df_null_pivot2 = df_null.pivot_table(index='income_type', columns='gender', values='debt', aggfunc='count')
df_null_pivot2

gender             F      M
income_type                
business       327.0  181.0
civil servant  112.0   35.0
employee       693.0  412.0
entrepreneur     NaN    1.0
retiree        352.0   61.0

**Kesimpulan**

Tidak ada pola yang terlihat jelas.

Dari eksplorasi yang dilakukan, saya menggabungakan kolom gender dan income_type. Dari sini bisa dilihat nasabah paling banyak adalah wanita dan pekerjaan paling banyak adalah employee, retiree, dan business.

Nilai yang hilang akan diganti dengan value yang sesuai atau mungkin dihapus.

Rencana singkat: Memeriksa nilai yang hilang dari data yang sudah difilter, menghilagkan nilai yang duplikat, memeriksa nilai unique, merubah nilai yang serupa tapi tak sama, dan merubah nilai yang merupakan salah input.

# Transformasi

In [14]:
df['education'].unique()

array(["bachelor's degree", 'secondary education', 'Secondary Education',
       'SECONDARY EDUCATION', "BACHELOR'S DEGREE", 'some college',
       'primary education', "Bachelor's Degree", 'SOME COLLEGE',
       'Some College', 'PRIMARY EDUCATION', 'Primary Education',
       'Graduate Degree', 'GRADUATE DEGREE', 'graduate degree'],
      dtype=object)

In [15]:
df['education'] = df['education'].str.lower()

In [16]:
df['education'].unique()

array(["bachelor's degree", 'secondary education', 'some college',
       'primary education', 'graduate degree'], dtype=object)

In [17]:
df['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

Data tersebut memiliki nilai yang unik seperti -1 dan 20. Seperti yang kita ketahui kalau anak tidak mungkin bernilai -1 dan nasabah yang memiliki anak berjumlah 20 itu sangat jarang bahkan bisa tidak ada. Sepertinya nilai tersebut terjadi kesalahan input. Mungkin -1 artinya adalah 1 dan 20 itu artinya 2.

Data tersebut bisa di-drop atau diganti nilainya menjadi nilai yang sesuai. Namun, di sini saya akan memilih untuk mengganti nilainya.

In [18]:
old_values = [-1, 20]
new_values = [1, 2]

df['children'] = df['children'].replace(old_values, new_values)

In [19]:
df['children'].value_counts()

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

In [20]:
df['days_employed'].unique()

array([-8437.67302776, -4024.80375385, -5623.42261023, ...,
       -2113.3468877 , -3112.4817052 , -1984.50758853])

In [21]:
percentage_days_employed = len(df[df['days_employed']<0])/len(df)
print ("{:.2%}".format(percentage_days_employed))

73.90%


In [22]:
df['days_employed'] = df['days_employed'].abs()

In [23]:
df['days_employed'].unique()

array([8437.67302776, 4024.80375385, 5623.42261023, ..., 2113.3468877 ,
       3112.4817052 , 1984.50758853])

In [24]:
df['dob_years'].unique()

array([42, 36, 33, 32, 53, 27, 43, 50, 35, 41, 40, 65, 54, 56, 26, 48, 24,
       21, 57, 67, 28, 63, 62, 47, 34, 68, 25, 31, 30, 20, 49, 37, 45, 61,
       64, 44, 52, 46, 23, 38, 39, 51,  0, 59, 29, 60, 55, 58, 71, 22, 73,
       66, 69, 19, 72, 70, 74, 75], dtype=int64)

In [25]:
df['dob_years'].value_counts(ascending=True)

75      1
74      6
73      8
19     14
72     33
20     51
71     58
70     65
69     85
68     99
0     101
21    111
67    167
66    183
22    183
65    194
23    254
24    264
64    265
63    269
62    352
61    355
25    357
60    377
26    408
55    443
59    444
51    448
53    459
57    460
58    461
46    475
54    479
47    480
52    484
56    487
27    493
45    497
28    503
49    508
32    510
43    513
50    514
37    537
48    538
30    540
29    545
44    547
36    555
31    560
39    573
33    581
42    597
38    598
34    603
41    607
40    609
35    617
Name: dob_years, dtype: int64

In [26]:
percatage_of_zero = len(df[df['dob_years']==0])/len(df) 
print ("{:.2%}".format(percatage_of_zero))

0.47%


Di kolom ini ada satu nilai yaitu nilai 0, yang mana umur nasabah itu tidak mungkin 0. Jumlahnya ada 101. Saya bisa menghapus data itu karena persentasi data hilang itu cuma 0.47% jadi tidak terlalu pengaruh terhadap analisis.

In [27]:
df = df.drop(df[df['dob_years']==0].index)

In [28]:
print(df[df['dob_years']==0])

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []


In [29]:
df['family_status'].value_counts()

married              12331
civil partnership     4156
unmarried             2797
divorced              1185
widow / widower        955
Name: family_status, dtype: int64

Tidak ada yang aneh dengan kolom family_status.

In [30]:
df['gender'].unique()

array(['F', 'M', 'XNA'], dtype=object)

Terdapat nilai XNA pada kolom gender. Ini mungkin bisa terjadi karena pemberian informasi yang salah ketika memasukkan data atau ada kesalahan sistem.

In [31]:
df[df['gender']=='XNA']

children  days_employed  dob_years     education  education_id  \
10701         0    2358.600502         24  some college             2   

           family_status  family_status_id gender income_type  debt  \
10701  civil partnership                 1    XNA    business     0   

       total_income          purpose  
10701     32624.825  buy real estate

Hanya terdapat satu nilai XNA. Bisa dihapus saja data dengan gender yang kurang jelas ini.

In [32]:
df = df.drop(df[df['gender']=='XNA'].index)

In [33]:
df['gender'].unique()

array(['F', 'M'], dtype=object)

Jenis gender XNA sudah hilang dari data.

In [34]:
df['income_type'].unique()

array(['employee', 'retiree', 'business', 'civil servant', 'unemployed',
       'entrepreneur', 'student', 'paternity / maternity leave'],
      dtype=object)

Tidak ada nilai yang bisa diatasi di sini. Nilai pada kolom ini sudah benar.

In [35]:
df[df.duplicated()]

children  days_employed  dob_years            education  education_id  \
2849          0            NaN         41  secondary education             1   
3290          0            NaN         58  secondary education             1   
4182          1            NaN         34    bachelor's degree             0   
4851          0            NaN         60  secondary education             1   
5557          0            NaN         58  secondary education             1   
...         ...            ...        ...                  ...           ...   
20702         0            NaN         64  secondary education             1   
21032         0            NaN         60  secondary education             1   
21132         0            NaN         47  secondary education             1   
21281         1            NaN         30    bachelor's degree             0   
21415         0            NaN         54  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
2849             married                 0      F    employee     0   
3290   civil partnership                 1      F     retiree     0   
4182   civil partnership                 1      F    employee     0   
4851   civil partnership                 1      F     retiree     0   
5557   civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
20702            married                 0      F     retiree     0   
21032            married                 0      F     retiree     0   
21132            married                 0      F    employee     0   
21281            married                 0      F    employee     0   
21415            married                 0      F     retiree     0   

       total_income                              purpose  
2849            NaN  purchase of the house for my family  
3290            NaN                    to have a wedding  
4182            NaN                     wedding ceremony  
4851            NaN                     wedding ceremony  
5557            NaN                    to have a wedding  
...             ...                                  ...  
20702           NaN              supplementary education  
21032           NaN                   to become educated  
21132           NaN                   housing renovation  
21281           NaN           buy commercial real estate  
21415           NaN                 housing transactions  

[71 rows x 12 columns]

In [36]:
df.duplicated().sum()

71

In [37]:
new_df = df.drop_duplicates().reset_index(drop=True)

In [38]:
new_df.duplicated().sum()

0

In [39]:
new_df.shape

(21352, 12)

In [40]:
new_df.tail(20)

children  days_employed  dob_years            education  education_id  \
21332         0  338904.866406         53  secondary education             1   
21333         1    1556.249906         33    bachelor's degree             0   
21334         1      79.832064         32  secondary education             1   
21335         0  386497.714078         62  secondary education             1   
21336         0  362161.054124         59    bachelor's degree             0   
21337         2            NaN         28  secondary education             1   
21338         0     612.569129         29    bachelor's degree             0   
21339         0     165.377752         26    bachelor's degree             0   
21340         0    1166.216789         35  secondary education             1   
21341         0     280.469996         27         some college             2   
21342         1     467.685130         28  secondary education             1   
21343         0     914.391429         42    bachelor's degree             0   
21344         0     404.679034         42    bachelor's degree             0   
21345         0  373995.710838         59  secondary education             1   
21346         1    2351.431934         37      graduate degree             4   
21347         1    4529.316663         43  secondary education             1   
21348         0  343937.404131         67  secondary education             1   
21349         1    2113.346888         38  secondary education             1   
21350         3    3112.481705         38  secondary education             1   
21351         2    1984.507589         40  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
21332  civil partnership                 1      M        retiree     0   
21333  civil partnership                 1      F       employee     0   
21334  civil partnership                 1      F  civil servant     0   
21335            married                 0      M        retiree     0   
21336            married                 0      M        retiree     0   
21337            married                 0      F       employee     0   
21338  civil partnership                 1      F       employee     1   
21339          unmarried                 4      M       business     0   
21340            married                 0      F       employee     0   
21341          unmarried                 4      M       business     0   
21342            married                 0      F       employee     1   
21343            married                 0      F       business     0   
21344  civil partnership                 1      F       business     0   
21345            married                 0      F        retiree     0   
21346           divorced                 3      M       employee     0   
21347  civil partnership                 1      F       business     0   
21348            married                 0      F        retiree     0   
21349  civil partnership                 1      M       employee     1   
21350            married                 0      M       employee     1   
21351            married                 0      F       employee     0   

       total_income                           purpose  
21332     12070.399                 to have a wedding  
21333     23286.719                  wedding ceremony  
21334     15708.845          second-hand car purchase  
21335     11622.175                          property  
21336     11684.650          real estate transactions  
21337           NaN                      car purchase  
21338     22410.956   buying property for renting out  
21339     23568.233  to get a supplementary education  
21340     40157.783             purchase of the house  
21341     56958.145               building a property  
21342     17517.812                to become educated  
21343     51649.244          purchase of my own house  
21344     28489.529                 buying 

Pada dataset ini kita sudah menghapus 173 data. Mulai dari data duplikat dan menghapus nilai yang unik. Selain itu data yang baru ini sudah diubah nilai yang ada kesalahannya di kolom children, days_employed, dan dob_years. Presentase data yang diubah sebesar **0.8%**, karena memang pada data ini untuk data yang memiliki kecatatan lebih sedikit kalau dibandingkan dengan data yang nilainya hilang yaitu sebesar **10%**.

# Bekerja dengan Nilai yang Hilang

In [41]:
education = new_df[['education', 'education_id']]
education = education.drop_duplicates()
education

education  education_id
0       bachelor's degree             0
1     secondary education             1
13           some college             2
31      primary education             3
2946      graduate degree             4

In [42]:
education_dict = dict(zip(education['education'], education['education_id']))
education_dict

{"bachelor's degree": 0,
 'secondary education': 1,
 'some college': 2,
 'primary education': 3,
 'graduate degree': 4}

In [43]:
family = new_df[['family_status', 'family_status_id']]
family = family.drop_duplicates()
family

family_status  family_status_id
0             married                 0
4   civil partnership                 1
18    widow / widower                 2
19           divorced                 3
24          unmarried                 4

In [44]:
family_dict = dict(zip(family['family_status'], family['family_status_id']))
family_dict

{'married': 0,
 'civil partnership': 1,
 'widow / widower': 2,
 'divorced': 3,
 'unmarried': 4}

Karena data yang diolah cukup banyak dan mungkin akan memakan cukup banyak waktu juga kalau saya mengfilter berdasarkan purpose. Jadi, kita perlu mengfilter lewat id.

Karena kategori yang memiliki id itu ada dua yaitu family_status dan education, maka kita perlu membuat dictionary dari kedua kategori tersebut.

## Nilai Hilang di total_income

Kolom yang perlu ditangani itu ada dua, yang pertama ada total_income dan yang kedua adalah days_employed. Untuk kolom di total_income kita perlu mengisi nilai yang hilang dengan nilai median atau nilai tengah. Sedangkan untuk kolom days_employed kita mengisinya dengan nilai rata-rata.

In [45]:
new_df['dob_years'].sort_values().unique()

array([19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35,
       36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52,
       53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69,
       70, 71, 72, 73, 74, 75], dtype=int64)

In [46]:
def age_category(age):
    if age < 30:
        return '19-29'
    elif age < 40:
        return '30-39'
    elif age < 50:
        return '40-49'
    elif age < 60:
        return '50-59'
    elif age < 70:
        return '60-69'
    elif age <= 75:
        return '70-75'

In [47]:
age_category(35)

'30-39'

In [48]:
new_df['age_category'] = new_df['dob_years'].apply(age_category)

In [49]:
new_df[['age_category']]

age_category
0            40-49
1            30-39
2            30-39
3            30-39
4            50-59
...            ...
21347        40-49
21348        60-69
21349        30-39
21350        30-39
21351        40-49

[21352 rows x 1 columns]

Biasanya faktor-faktor yang bergantung pada pendapatan adalah usia, pendidikan, keluarga, pekerjaan. Untuk usia, semakin lama seseorang bekerja, maka semakin tinggi gaji yang didapat dikarenakan pengalamannya. Untuk pendidikan, biasanya semakin tinggi pendidikan seseorang, maka gaji atau pangkatnya semakin tinggi. Untuk keluarga, jika seseorang sudah berkeluarga maka akan diberikan tunjangan tergantung jumlah anggota keluarga yang dipunya. Untuk pekerjaan, setiap pekerjaan memiliki pendapatan yang berbeda-beda.

In [50]:
df_without_null = new_df[new_df['days_employed'].notnull()]
df_without_null.head(10)

children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   
5         0     926.185831         27    bachelor's degree             0   
6         0    2879.202052         43    bachelor's degree             0   
7         0     152.779569         50  secondary education             1   
8         2    6929.865299         35    bachelor's degree             0   
9         0    2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose age_category  
0                purchase of the house        40-49  
1                         car purchase        30-39  
2                purchase of the house        30-39  
3              supplementary education        30-39  
4                    to have a wedding        50-59  
5                purchase of the house        19-29  
6                 housing transactions        40-49  
7                            education        50-59  
8                     having a wedding        30-39  
9  purchase of the house for my family        40-49

In [51]:
test_age_mean = df_without_null.pivot_table(index='age_category', values='total_income', aggfunc='mean')
test_age_mean.sort_values(by='total_income', ascending=False)

total_income
age_category              
40-49         28551.375635
30-39         28312.479963
50-59         25811.700327
19-29         25531.501098
60-69         23242.812818
70-75         20125.658331

In [52]:
test_age_median = df_without_null.pivot_table(index='age_category', values='total_income', aggfunc='median')
test_age_median.sort_values(by='total_income', ascending=False)

total_income
age_category              
40-49           24764.2290
30-39           24667.5280
19-29           22735.9110
50-59           22203.0745
60-69           19817.4400
70-75           18751.3240

In [53]:
test_education_mean = df_without_null.pivot_table(index='age_category', columns='education', values='total_income', aggfunc='mean')
test_education_mean

education     bachelor's degree  graduate degree  primary education  \
age_category                                                          
19-29              29395.106109              NaN       27695.271520   
30-39              34225.243752       18187.3015       21979.584515   
40-49              35687.086166       31771.3210       23618.267667   
50-59              33622.674858       42945.7940       17797.322623   
60-69              30329.905325       28334.2150       18710.592883   
70-75              26173.068696              NaN       18892.886000   

education     secondary education  some college  
age_category                                     
19-29                23379.052855  25265.724779  
30-39                25666.783012  31991.246531  
40-49                26193.975154  33703.486528  
50-59                24057.271660  27623.492353  
60-69                21691.515498  30476.594957  
70-75                19245.043953  13917.989667

In [54]:
test_education_median = df_without_null.pivot_table(index='age_category', columns='education', values='total_income', aggfunc='median')
test_education_median

education     bachelor's degree  graduate degree  primary education  \
age_category                                                          
19-29                25956.1640              NaN         25488.9160   
30-39                28794.9310       18187.3015         19542.3265   
40-49                30282.3330       31771.3210         21511.5635   
50-59                28152.1765       42945.7940         16922.6250   
60-69                25222.3445       28334.2150         17657.4995   
70-75                25497.3920              NaN         15013.5050   

education     secondary education  some college  
age_category                                     
19-29                   21114.762     22655.467  
30-39                   22912.993     28463.439  
40-49                   22973.258     29323.673  
50-59                   21245.482     21132.309  
60-69                   18873.764     28178.917  
70-75                   18508.577     14479.193

In [55]:
test_family_mean = df_without_null.pivot_table(index='age_category', columns='family_status', values='total_income', aggfunc='mean')
test_family_mean

family_status  civil partnership      divorced       married     unmarried  \
age_category                                                                 
19-29               25567.116944  25073.832144  25536.319673  25553.297719   
30-39               27714.148395  29115.091847  28573.954796  27739.154427   
40-49               28956.763892  28275.677651  28440.743835  29572.001536   
50-59               25009.327522  25679.409055  26298.481978  27466.250638   
60-69               23044.254153  25911.107331  22982.894501  23566.604574   
70-75               20728.343107  22049.877600  20000.835831  21892.799385   

family_status  widow / widower  
age_category                    
19-29             22915.450000  
30-39             27781.924289  
40-49             24929.826568  
50-59             22177.332934  
60-69             23004.881780  
70-75             19048.421270

In [56]:
test_family_median = df_without_null.pivot_table(index='age_category', columns='family_status',values='total_income', aggfunc='median')
test_family_median

family_status  civil partnership    divorced     married   unmarried  \
age_category                                                           
19-29                 22468.1820  22276.4620  22799.2580  22895.1195   
30-39                 23975.6400  24960.6380  25027.4530  23752.9560   
40-49                 24966.6930  25138.1490  24668.6130  25788.1610   
50-59                 21628.2120  22212.5010  22596.0325  22775.1220   
60-69                 19905.6490  21057.2675  19877.7880  19676.3690   
70-75                 19934.7915  22101.5060  19163.3830  16860.3290   

family_status  widow / widower  
age_category                    
19-29               22915.4500  
30-39               26565.3295  
40-49               22723.8610  
50-59               20389.9210  
60-69               19301.9280  
70-75               17971.7290

In [57]:
test_job_median = df_without_null.pivot_table(index='age_category', columns='income_type',values='total_income', aggfunc='mean')
test_job_median

income_type       business  civil servant      employee  entrepreneur  \
age_category                                                            
19-29         28685.453876   25145.840248  24099.859203     79866.103   
30-39         33145.949183   27921.836553  26191.716908           NaN   
40-49         33989.836505   28568.272491  26193.926281           NaN   
50-59         32385.032725   25838.105730  26073.759931           NaN   
60-69         32494.918350   29305.166039  27307.606610           NaN   
70-75         27766.307200   32189.795667  26672.382429           NaN   

income_type   paternity / maternity leave       retiree   student  unemployed  
age_category                                                                   
19-29                                 NaN  14888.651857  15712.26         NaN  
30-39                            8612.661  23122.709862       NaN    9593.119  
40-49                                 NaN  27020.126339       NaN   32435.602  
50-59                                 NaN  22221.765833       NaN         NaN  
60-69                                 NaN  21544.426743       NaN         NaN  
70-75                                 NaN  18994.044264       NaN         NaN

In [58]:
test_job_median = df_without_null.pivot_table(index='age_category', columns='income_type',values='total_income', aggfunc='median')
test_job_median

income_type     business  civil servant   employee  entrepreneur  \
age_category                                                       
19-29         25542.5850     23504.5105  21537.479     79866.103   
30-39         28765.5935     24662.7440  23218.803           NaN   
40-49         28698.4665     24890.7590  23108.150           NaN   
50-59         27299.6345     23136.3020  22547.831           NaN   
60-69         29171.9890     23390.0570  23316.965           NaN   
70-75         28138.8950     24525.2240  24660.901           NaN   

income_type   paternity / maternity leave     retiree   student  unemployed  
age_category                                                                 
19-29                                 NaN  12807.0710  15712.26         NaN  
30-39                            8612.661  18735.7160       NaN    9593.119  
40-49                                 NaN  22498.7080       NaN   32435.602  
50-59                                 NaN  19526.8120       NaN         NaN  
60-69                                 NaN  18446.1435       NaN         NaN  
70-75                                 NaN  17650.4660       NaN         NaN

Karakteristik yang paling menentukan adalah pekerjaan dan usia pekerja. Karena ketika seorang mendapatkan pekerjaan yang bagus ditambah usianya ada yang senior dan junior, maka itu faktor utama yang memengaruhi pendapatan. Sebenarnya faktor lain juga berpengaruh tapi tidak sebepengaruh usia dan jenis pekerjaan. Kita ambil contoh family_status, seseorang bisa mendapatkan gaji yang lebih tinggi ketika sudah menikah dan memiliki anak karena mendapatkan tunjangan, tapi ada seseorang yang mendapatkan gaji yang sama bahkan lebih padahal dia masih muda dan blm menikah namun jenis pekerjaannya berbeda.

Dalam kasus ini, saya akan menggunakan median untuk mengisi nilai yang hilang. Dikarenakan outlier yg signifikan untuk setiap orang. Akan cukup adil jika diisi berdasarkan median.

In [59]:
def get_median(income_type, age_category):
    try:
        return test_job_median[income_type][age_category]
    except:
        return "error"

In [60]:
get_median('business', '19-29')

25542.585

In [61]:
new_df['median_income'] = new_df.apply(lambda x: get_median(x['income_type'], x['age_category']), axis=1)

In [62]:
new_df.head(10)

children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   
5         0     926.185831         27    bachelor's degree             0   
6         0    2879.202052         43    bachelor's degree             0   
7         0     152.779569         50  secondary education             1   
8         2    6929.865299         35    bachelor's degree             0   
9         0    2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose age_category  median_income  
0                purchase of the house        40-49     23108.1500  
1                         car purchase        30-39     23218.8030  
2                purchase of the house        30-39     23218.8030  
3              supplementary education        30-39     23218.8030  
4                    to have a wedding        50-59     19526.8120  
5                purchase of the house        19-29     25542.5850  
6                 housing transactions        40-49     28698.4665  
7                            education        50-59     22547.8310  
8                     having a wedding        30-39     23218.8030  
9  purchase of the house for my family        40-49     23108.1500

In [63]:
new_df['total_income'] = new_df['total_income'].fillna(new_df['median_income'])

In [64]:
new_df.isnull().sum()

children               0
days_employed       2093
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income           1
purpose                0
age_category           0
median_income          1
dtype: int64

Terdapat kesalahan dalam kolom total_income dan median_values yang mana pada kedua kolom tersebut tidak dapat diisi menggunakan fillna(). Jadi kita isi secara manual menggunakan nilai rata-rata di usia 50-59.

In [65]:
new_df.at[5907, 'total_income'] = 25811.700327
new_df.at[5907, 'median_income'] = 25811.700327

In [66]:
print(new_df.loc[5907, 'total_income'])
print(new_df.loc[5907, 'median_income'])

25811.700327
25811.700327


In [67]:
new_df.isna().sum()

children               0
days_employed       2093
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income           0
purpose                0
age_category           0
median_income          0
dtype: int64

In [68]:
new_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21352 entries, 0 to 21351
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21352 non-null  int64  
 1   days_employed     19259 non-null  float64
 2   dob_years         21352 non-null  int64  
 3   education         21352 non-null  object 
 4   education_id      21352 non-null  int64  
 5   family_status     21352 non-null  object 
 6   family_status_id  21352 non-null  int64  
 7   gender            21352 non-null  object 
 8   income_type       21352 non-null  object 
 9   debt              21352 non-null  int64  
 10  total_income      21352 non-null  float64
 11  purpose           21352 non-null  object 
 12  age_category      21352 non-null  object 
 13  median_income     21352 non-null  float64
dtypes: float64(3), int64(5), object(6)
memory usage: 2.3+ MB


## Nilai Hilang di days_employed

In [69]:
days_employed_median = new_df.pivot_table(index='age_category', columns='income_type', values='days_employed', aggfunc='median')
days_employed_median

income_type      business  civil servant     employee  entrepreneur  \
age_category                                                          
19-29          906.647054    1362.645769  1008.784193    520.848083   
30-39         1526.857305    2591.886944  1527.637665           NaN   
40-49         1896.569279    3551.609375  1904.197739           NaN   
50-59         2003.075639    3822.891349  2235.163052           NaN   
60-69         2358.275817    3339.663548  2669.706829           NaN   
70-75         3095.344969    1678.969771  1504.924191           NaN   

income_type   paternity / maternity leave        retiree     student  \
age_category                                                           
19-29                                 NaN  364348.197352  578.751554   
30-39                         3296.759962  365336.560325         NaN   
40-49                                 NaN  367477.402016         NaN   
50-59                                 NaN  364343.088140         NaN   
60-69                                 NaN  365484.090620         NaN   
70-75                                 NaN  366503.463685         NaN   

income_type      unemployed  
age_category                 
19-29                   NaN  
30-39         337524.466835  
40-49         395302.838654  
50-59                   NaN  
60-69                   NaN  
70-75                   NaN

In [70]:
days_employed_mean = new_df.pivot_table(index='age_category', columns='income_type', values='days_employed', aggfunc='mean')
days_employed_mean

income_type      business  civil servant     employee  entrepreneur  \
age_category                                                          
19-29         1128.501894    1595.684495  1202.858627    520.848083   
30-39         1854.594279    2764.437287  2010.804449           NaN   
40-49         2477.016422    4032.418745  2673.300396           NaN   
50-59         2795.567404    4825.214157  3254.446979           NaN   
60-69         3477.081000    4433.159823  3916.746275           NaN   
70-75         5292.598849    2811.783573  3310.691322           NaN   

income_type   paternity / maternity leave        retiree     student  \
age_category                                                           
19-29                                 NaN  360706.968329  578.751554   
30-39                         3296.759962  365461.247788         NaN   
40-49                                 NaN  366963.446993         NaN   
50-59                                 NaN  364611.037392         NaN   
60-69                                 NaN  365158.698699         NaN   
70-75                                 NaN  366046.629499         NaN   

income_type      unemployed  
age_category                 
19-29                   NaN  
30-39         337524.466835  
40-49         395302.838654  
50-59                   NaN  
60-69                   NaN  
70-75                   NaN

Untuk kolom days_employed saya menggunakan mean atau nilai rata-rata. karena outlier dari data tersebut tidak terlalu signifikan. Sehigga kita bisa mengisi nilai yang kosong berdasarkan nilai rata-rata dari umur dan posisi kerja.

In [71]:
def get_mean_days(income_type, age_category):
    return days_employed_mean[income_type][age_category]

In [72]:
new_df['mean_days'] = new_df.apply(lambda x: get_mean_days(x['income_type'], x['age_category']), axis=1)

In [73]:
get_mean_days('business', '40-49')

2477.0164217983383

In [74]:
new_df.head(10)

children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   
5         0     926.185831         27    bachelor's degree             0   
6         0    2879.202052         43    bachelor's degree             0   
7         0     152.779569         50  secondary education             1   
8         2    6929.865299         35    bachelor's degree             0   
9         0    2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose age_category  median_income  \
0                purchase of the house        40-49     23108.1500   
1                         car purchase        30-39     23218.8030   
2                purchase of the house        30-39     23218.8030   
3              supplementary education        30-39     23218.8030   
4                    to have a wedding        50-59     19526.8120   
5                purchase of the house        19-29     25542.5850   
6                 housing transactions        40-49     28698.4665   
7                            education        50-59     22547.8310   
8                     having a wedding        30-39     23218.8030   
9  purchase of the house for my family        40-49     23108.1500   

       mean_days  
0    2673.300396  
1    2010.804449  
2    2010.804449  
3    2010.804449  
4  364611.037392  
5    1128.501894  
6    2477.016422  
7    3254.446979  
8    2010.804449  
9    2673.300396

In [75]:
new_df['days_employed'] = new_df['days_employed'].fillna(new_df['mean_days'])

In [76]:
new_df.isnull().sum()

children            0
days_employed       1
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
age_category        0
median_income       0
mean_days           1
dtype: int64

In [77]:
new_df[new_df['days_employed'].isna()]

children  days_employed  dob_years          education  education_id  \
5907         0            NaN         58  bachelor's degree             0   

     family_status  family_status_id gender   income_type  debt  total_income  \
5907       married                 0      M  entrepreneur     0  25811.700327   

                          purpose age_category  median_income  mean_days  
5907  buy residential real estate        50-59   25811.700327        NaN

Kesalahan yang sama namun di kolom yang berbeda. Diisi dengan cara manual dan dengan nilai berdasarkan kategori umur.

In [78]:
age_mean = new_df.pivot_table(index='age_category', values='days_employed', aggfunc='mean')
age_mean

days_employed
age_category               
19-29           2114.390998
30-39           4268.104511
40-49          12466.515797
50-59         133488.932188
60-69         282851.863348
70-75         318946.710089

In [79]:
new_df.at[5907, 'days_employed'] = 133488.932188
new_df.at[5907, 'mean_days'] = 133488.932188

In [80]:
new_df.isnull().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
age_category        0
median_income       0
mean_days           0
dtype: int64

Sudah tidak ada lagi nilai yang hilang pada data ini.

## Pengkategorian Data

In [81]:
new_df[['purpose']]

purpose
0        purchase of the house
1                 car purchase
2        purchase of the house
3      supplementary education
4            to have a wedding
...                        ...
21347     housing transactions
21348        purchase of a car
21349                 property
21350        buying my own car
21351             to buy a car

[21352 rows x 1 columns]

In [82]:
new_df['purpose'].sort_values().unique()

array(['building a property', 'building a real estate',
       'buy commercial real estate', 'buy real estate',
       'buy residential real estate', 'buying a second-hand car',
       'buying my own car', 'buying property for renting out', 'car',
       'car purchase', 'cars', 'construction of own property',
       'education', 'getting an education', 'getting higher education',
       'going to university', 'having a wedding', 'housing',
       'housing renovation', 'housing transactions', 'profile education',
       'property', 'purchase of a car', 'purchase of my own house',
       'purchase of the house', 'purchase of the house for my family',
       'real estate transactions', 'second-hand car purchase',
       'supplementary education', 'to become educated', 'to buy a car',
       'to get a supplementary education', 'to have a wedding',
       'to own a car', 'transactions with commercial real estate',
       'transactions with my real estate', 'university education',
       'we

Dari nilai unik di atas, saya bisa mendapatkan topik umum yaitu: property, real estate, car, education, wedding, dan house

In [95]:
# sample_dict = {
#     'building a property':'property',
#     'buying property for renting out':'property',
#     'construction of own property':'property',
#     'property':'property',
#     'building a real estate':'real estate',
#     'buy commercial real estate':'real estate',
#     'buy real estate':'real estate',
#     'buy residential real estate':'real estate',
#     'real estate transactions':'real estate',
#     'transactions with commercial real estate':'real estate',
#     'transactions with my real estate':'real estate',
#     'buying a second-hand car':'car',
#     'buying my own car':'car',
#     'car':'car',
#     'car purchase':'car',
#     'cars':'car',
#     'purchase of a car':'car',
#     'second-hand car purchase':'car',
#     'to buy a car':'car',
#     'to own a car':'car',
#     'education':'education',
#     'getting an education':'education',
#     'getting higher education':'education',
#     'profile education':'education',
#     'supplementary education':'education',
#     'to get a supplementary education':'education',
#     'university education':'education',
#     'going to university':'education',
#     'to become educated':'education',
#     'having a wedding':'wedding',
#     'to have a wedding':'wedding',
#     'wedding ceremony':'wedding',
#     'housing':'house',
#     'housing renovation':'house',
#     'housing transactions':'house',
#     'purchase of my own house':'house',
#     'purchase of the house':'house',
#     'purchase of the house for my family':'house'
# }
def purpose_common(purpose):
    if 'property' in purpose:
        return 'property'
    elif 'estate' in purpose:
        return 'property'
    elif 'hous' in purpose:
        return 'property'
    elif 'car' in purpose:
        return 'car'
    elif 'educ' in purpose:
        return 'education'
    elif 'univ' in purpose:
        return 'education'
    elif 'wedd' in purpose:
        return 'wedding'

In [96]:
new_df['purpose_categorize'] = new_df['purpose'].apply(purpose_common)

In [97]:
new_df['purpose_categorize'].value_counts()

property     10763
car           4284
education     3995
wedding       2310
Name: purpose_categorize, dtype: int64

In [98]:
new_df[['total_income']]

total_income
0         40620.102
1         17932.802
2         23341.752
3         42820.568
4         25378.572
...             ...
21347     35966.698
21348     24959.969
21349     14347.610
21350     39054.888
21351     13127.587

[21352 rows x 1 columns]

In [99]:
new_df['total_income'].describe()

count     21352.000000
mean      26464.393364
std       15725.852146
min        3306.762000
25%       17222.623000
50%       23136.302000
75%       31321.653000
max      362496.645000
Name: total_income, dtype: float64

In [100]:
def income_categorize(income):
    if income <= 50000:
        return "low"
    elif income <= 100000:
        return "middle"
    elif income > 100000:
        return "high"

In [101]:
new_df['income_categorize'] = new_df['total_income'].apply(income_categorize)

In [102]:
new_df['income_categorize'].value_counts()

low       20034
middle     1219
high         99
Name: income_categorize, dtype: int64

## Memeriksa Hipotesis

**Apakah terdapat korelasi antara memiliki anak dengan probabilitas melakukan gagal bayar pinjaman?**

In [103]:
pivot_table_children = new_df.pivot_table(index='children', columns='debt' ,values='total_income', aggfunc='count')
pivot_table_children[1] = pivot_table_children[1].fillna(0)
pivot_table_children['percentage_debt'] = pivot_table_children[1]/(pivot_table_children[1]+pivot_table_children[0])
pivot_table_children['percentage_debt'] = pivot_table_children['percentage_debt'].apply("{:,.2%}".format)
pivot_table_children

debt            0       1 percentage_debt
children                                 
0         12963.0  1058.0           7.55%
1          4397.0   442.0           9.13%
2          1912.0   202.0           9.56%
3           301.0    27.0           8.23%
4            37.0     4.0           9.76%
5             9.0     0.0           0.00%

Untuk nasabah yang tidak memiliki anak memiliki presentasi tidak bayar sangat rendah dibandingkan yang lain. Hal ini mungkin bisa terjadi karena masih belum banyak tanggungan seperti tanggungan anak. Sehingga bisa melunasi pinjaman dengan lancar. Namun, hal yang tidak terduga adalah nasabah dengan anak 5 yang bisa melunasi hutang. Apakah ini disebabkan karena pendapatan mereka yang besar?

**Apakah terdapat korelasi antara status keluarga dengan probabilitas melakukan gagal bayar pinjaman?**

In [104]:
pivot_table_family= new_df.pivot_table(index='family_status', columns='debt' ,values='total_income', aggfunc='count')
pivot_table_family['percentage_debt'] = pivot_table_family[1]/(pivot_table_family[1]+pivot_table_family[0])
pivot_table_family['percentage_debt'] = pivot_table_family['percentage_debt'].apply("{:,.2%}".format)
pivot_table_family

debt                   0    1 percentage_debt
family_status                                
civil partnership   3743  386           9.35%
divorced            1100   85           7.17%
married            11363  927           7.54%
unmarried           2521  273           9.77%
widow / widower      892   62           6.50%

Untuk nasabah yang sudah janda atau duda, memiliki persentase lebih rendah untuk tidak bayar. Hal ini mungkin bisa terjadi karena sebagian besar dari mereka itu sudah hidup sendiri dan tidak memiliki tanggungan untuk keluarga atau anak. Yang paling tertinggi adalah nasabah dengan status unmarried atau belum menikah. 

**Bagaimana tujuan kredit memengaruhi persentase gagal bayar?**

In [105]:
pivot_table_credit = new_df.pivot_table(index='purpose_categorize', columns='debt' ,values='total_income', aggfunc='count')
pivot_table_credit['percentage_debt'] = pivot_table_credit[1]/(pivot_table_credit[1]+pivot_table_credit[0])
pivot_table_credit['percentage_debt'] = pivot_table_credit['percentage_debt'].apply("{:,.2%}".format)
pivot_table_credit

debt                   0    1 percentage_debt
purpose_categorize                           
car                 3884  400           9.34%
education           3625  370           9.26%
property            9984  779           7.24%
wedding             2126  184           7.97%

Untuk nasabah yang ingin membeli mobil memiliki presentase gagal bayar yang tertinggi. Ini mungkin bisa terjadi karena dalam membeli mobil bukan hanya sekedar melunasi pembayarannya tapi juga diperlukan biaya pajak dan lain-lain. Sedangkan untuk yang terendah terdapat pada nasabah yang membeli property. Dalam membeli property, bisa kita mencicil property itu setiap bulan atau setiap tahun dan bisa menjalankan bisnis ketika kita sudah mempunyai property. Sehingga, membuat kita bisa mudah dalam melunasi peminjaman.

## Kesimpulan Umum

Analisis dimulai dengan kita membuat hipotesis. Hipotesis ini meliputi:
1. Seseorang gagal dalam membayar pinjaman karena ada pengeluaran lain untuk istri dan anaknya
2. Seseorang gagal dalam membayar pinjaman karena pendapatannya yang ternyata tidak cukup.
3. Perbedaan tujuan dalam melakukan pinjaman bisa menjadi faktor seseorang gagal dalam melakukan pinjaman.

Sebelum menguji hipotesis tersebut, kita terlebih dahulu melakukan pra-pemrosesan pada data, diantaranya:
1. Mengubah nilai pada kolom education menjadi huruf kecil semua.
2. Merubah nilai yang salah pada kolom children.
3. Merubah nilai negatif pada kolom days_employed menjadi nilai positif.
4. Menghapus nilai 0 pada komo dob_years.
5. Menghapus nilai unik XNA pada kolom gender.
6. Menghapus nilai duplikat.

Setelah semua data sudah bersih, maka dilanjut dengan mengisi nilai yang hilang:
- Kolom total_income:
1. Mengkategorikan kolom dob_years menjadi rentang tahun.
2. Setelah itu saya mencari tahu kategori apa yang sangat memengaruhi pendapatan seseorang. Ternyata, pekerjaan dan umur merupakan pengaruh yang besar dalam pendapatan seseorang.
3. Setelah itu kita mencari nilai mana yang lebih cocok untuk mengisi nilai yang hilang, mean atau median.
4. Ternyata nilai median merupakan nilai yang cocok dalam mengisi nilai yang hilang. Dikarenakan outlier yang begitu besar.
- Kolom dob_years:
1. Langkah hampir sama dengan total_income.
2. Di sini saya menggunakan nilai rata-rata karena oultlier yang tidak begitu besar. Sehingga adil ketika saya mengisi nilai yang hilang dengan median.

Pada pengisian nilai yang hilang ada satu nilai yang tidak terisi yaitu pada baris ke-5907. Jadi, saya mengisi nilainya sesuai dengan nilai median dan mean sesuai umurnya yaitu 50-59.

Setelah itu saya mengkategorikan kolom income_type dan purpose agar mudah untuk dianalisa.

Pengujian Hipotesis pun dimulai dan menghasilkan hal-hal berikut:
1. Untuk nasabah yang tidak memiliki anak memiliki presentasi tidak bayar sangat rendah dibandingkan yang lain. Hal ini mungkin bisa terjadi karena masih belum banyak tanggungan seperti tanggungan anak. Sehingga bisa melunasi pinjaman dengan lancar. Namun, hal yang tidak terduga adalah nasabah dengan anak 5 yang bisa melunasi hutang. Apakah ini disebabkan karena pendapatan mereka yang besar?
2. Untuk nasabah yang sudah janda atau duda, memiliki persentase lebih rendah untuk tidak bayar. Hal ini mungkin bisa terjadi karena sebagian besar dari mereka itu sudah hidup sendiri dan tidak memiliki tanggungan untuk keluarga atau anak. Yang paling tertinggi adalah nasabah dengan status unmarried atau belum menikah.
3. Untuk nasabah yang ingin membeli mobil memiliki presentase gagal bayar yang tertinggi. Ini mungkin bisa terjadi karena dalam membeli mobil bukan hanya sekedar melunasi pembayarannya tapi juga diperlukan biaya pajak dan lain-lain. Sedangkan untuk yang terendah terdapat pada nasabah yang membeli rumah. Dalam membeli rumah, bisa kita mencicil rumah itu setiap bulan atau setiap tahun. Sehingga, membuat kita bisa mudah dalam melunasi peminjaman.

Namun, dari semua tabel di atas dapat ditarik kesimpulan kalau presentasi tertinggi orang yang tidak membayar pinjaman yaitu orang yang belum menikah. Hal ini mungkin bisa terjadi karena alasan dia meminjam uang. Entah itu untuk membeli mobil atau hal yang lain.